src/Functions/generateSnowflakeID.cpp

```diff
@@ -0,0 +1,195 @@
+├─┼                 ┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┤
+|                   |     machine_id    |    machine_seq_num    |
+├─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┼─┤
+
+- The first 41 (+ 1 top zero bit) bits is the timestamp (millisecond since Unix epoch 1 Jan 1970)
+- The middle 10 bits are the machine ID
+- The last 12 bits are a counter to disambiguate multiple snowflakeIDs generated within the same millisecond by differen processes
+*/
+
+constexpr auto timestamp_bits_count = 41;
+constexpr auto machine_id_bits_count = 10;
+constexpr auto machine_seq_num_bits_count = 12;
+
+constexpr int64_t timestamp_mask = ((1LL << timestamp_bits_count) - 1) << (machine_id_bits_count + machine_seq_num_bits_count);
+constexpr int64_t machine_id_mask = ((1LL << machine_id_bits_count) - 1) << machine_seq_num_bits_count;
+constexpr int64_t machine_seq_num_mask = (1LL << machine_seq_num_bits_count) - 1;
+constexpr int64_t max_machine_seq_num = machine_seq_num_mask;
+
+Int64 getMachineID()
+{
+    UUID server_uuid = ServerUUID::get();
+    /// hash into 64 bits
+    UInt64 hi = UUIDHelpers::getHighBytes(server_uuid);
+    UInt64 lo = UUIDHelpers::getLowBytes(server_uuid);
+    return ((hi * 11) ^ (lo * 17)) & machine_id_mask;
+}
+
+Int64 getTimestamp()
+{
+    auto now = std::chrono::system_clock::now();
+    auto ticks_since_epoch = std::chrono::duration_cast<std::chrono::milliseconds>(now.time_since_epoch()).count();
+    return ticks_since_epoch & ((1LL << timestamp_bits_count) - 1);
+}
+
+}
+
+class FunctionSnowflakeID : public IFunction
+{
+private:
+    mutable std::atomic<Int64> lowest_available_snowflake_id = 0; /// atomic to avoid a mutex
+
+public:
+    static constexpr auto name = "generateSnowflakeID";
+    static FunctionPtr create(ContextPtr /*context*/) { return std::make_shared<FunctionSnowflakeID>(); }
+
+    String getName() const override { return name; }
+    size_t getNumberOfArguments() const override { return 0; }
+    bool isDeterministic() const override { return false; }
+    bool isDeterministicInScopeOfQuery() const override { return false; }
+    bool useDefaultImplementationForNulls() const override { return false; }
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return false; }
+    bool isVariadic() const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        if (!arguments.empty()) {
+            throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH,
+                "Number of arguments for function {} doesn't match: passed {}, should be 0.",
+                getName(), arguments.size());
+        }
+        return std::make_shared<DataTypeInt64>();
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & /*arguments*/, const DataTypePtr &, size_t input_rows_count) const override
```

Comment:
`generateUUIDv7` provides three alternatives for monotonicy within the counter / sequence field (global, thread-local, none): https://clickhouse.com/docs/en/sql-reference/functions/uuid-functions#generateUUIDv7

Consider if similarly three different variants make sense for `generateSnowflakeID`.

Reply:
It looks like the thread-local policy as it´s implemented now can cause SnowflakeID collisions when SnowflakeIDs are being generated within different threads simultaneously. In UUIDv7 there are a random part `rand_b` (having 32 bits size) and a random start value of a counter (having 42 bits size) which effectively minimize collision probability in the cases of not global policies, but there is no such random parts in SnowflakeID, so the initial implementation with synchronized retrieval a monotonic range from a global pool is the only one that should remain.
Also comparing to UUIDv7 generators the "correct" global policy would be taking counter values one by one for each row instead of taking a consecutive range for all the rows, and i don´t think that such "correct" implementation is really needed for SnowflakeIDs.

Reply:
I agree.